ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java

```diff
@@ -0,0 +1,138 @@
+ */
+package com.github.ambry.cloud.azure;
+
+import com.github.ambry.cloud.CloudDestination;
+import com.github.ambry.commons.BlobId;
+import com.microsoft.azure.storage.CloudStorageAccount;
+import com.microsoft.azure.storage.OperationContext;
+import com.microsoft.azure.storage.blob.BlobContainerPublicAccessType;
+import com.microsoft.azure.storage.blob.BlobRequestOptions;
+import com.microsoft.azure.storage.blob.CloudBlobClient;
+import com.microsoft.azure.storage.blob.CloudBlobContainer;
+import com.microsoft.azure.storage.blob.CloudBlockBlob;
+import java.io.InputStream;
+import java.util.Objects;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Implementation of {@link CloudDestination} that interacts with Azure Blob Storage service.
+ */
+class AzureCloudDestination implements CloudDestination {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AzureCloudDestination.class);
+  private final CloudStorageAccount azureAccount;
+  private final CloudBlobClient azureBlobClient;
+
+  /**
+   * Construct an Azure cloud destination from a container's replication config.
+   * @param configSpec the config spec to use.
+   * @throws Exception
+   */
+  AzureCloudDestination(String configSpec) throws Exception {
+    this(configSpec, null);
+  }
+
+  /**
+   * Construct an Azure cloud destination from a container's replication config and a {@link CloudStorageAccount} instance.
+   * @param configSpec the config spec to use.
+   * @param azureAccount the {@link CloudStorageAccount} to use.
+   * @throws Exception
+   */
+  AzureCloudDestination(String configSpec, CloudStorageAccount azureAccount) throws Exception {
+    if (azureAccount == null) {
+      azureAccount = CloudStorageAccount.parse(configSpec);
+    }
+    this.azureAccount = azureAccount;
+
+    // Create a blob client to interact with Blob storage
+    azureBlobClient = azureAccount.createCloudBlobClient();
+    LOGGER.info("Created Azure destination");
+  }
+
+  @Override
+  public boolean uploadBlob(BlobId blobId, long blobSize, InputStream blobInputStream) throws Exception {
+
+    Objects.requireNonNull(blobId, "BlobId cannot be null");
+    Objects.requireNonNull(blobInputStream, "Input stream cannot be null");
+
+    BlobRequestOptions options = null; // may want to set BlobEncryptionPolicy here
+    OperationContext opContext = null;
+    try {
+      CloudBlobContainer azureContainer = getContainer(blobId, true);
+      CloudBlockBlob azureBlob = azureContainer.getBlockBlobReference(blobId.getID());
+
+      if (azureBlob.exists()) {
+        LOGGER.debug("Skipping upload of blob {} as it already exists in Azure container {}.", blobId, azureContainer.getName());
+        return false;
+      }
+
+      azureBlob.upload(blobInputStream, blobSize, null, options, opContext);
+      LOGGER.debug("Uploaded blob {} to Azure container {}.", blobId, azureContainer.getName());
+      return true;
+    } catch (Exception e) {
```

Comment:
Still question about the Exception. 
Looks like we can use `throw StorageException and URISyntaxException` in line 67, which is also compatible with the interface. Do we need to cast them to `Exception` here?

Reply:
I agree about using generic Exception.  However, the interface CloudDestination's methods can't throw StorageException since that is an Azure class.  I could throw AmbryException, IOException, or make a new exception like CloudStoreException.  Any preference?

Reply:
All methods now throw new exception class CloudStorageException.